File: src/main/java/org/dukcode/ps/codetree/trail02/chapter04/lesson03/ChallengePaintingWhiteBlack.java
package org.dukcode.ps.codetree.trail02.chapter04.lesson03;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.StringTokenizer;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-painting-white-black/description">LINK</a>
 */
public class ChallengePaintingWhiteBlack {

  private static final int RANGE = 100_000;

  private static final int NONE = 0;
  private static final int WHITE = 1;
  private static final int BLACK = 2;
  private static final int GRAY = 3;

  private static BufferedReader br;
  private static BufferedWriter bw;
  private static StringTokenizer st;

  private static int[] currentColor;
  private static int[] numWhite;
  private static int[] numBlack;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    currentColor = new int[2 * RANGE + 1];
    numWhite = new int[2 * RANGE + 1];
    numBlack = new int[2 * RANGE + 1];

    int pos = 0;

    int minPos = 0;
    int maxPos = 0;
    int n = Integer.parseInt(br.readLine());
    for (int i = 0; i < n; i++) {
      st = new StringTokenizer(br.readLine());
      int cnt = Integer.parseInt(st.nextToken());
      int dir = st.nextToken().equals("L") ? -1 : 1;
      int color = dir == -1 ? WHITE : BLACK;

      for (int move = 0; move < cnt - 1; ++move) {
        paintColor(color, pos);
        pos += dir;
      }
      paintColor(color, pos);

      minPos = Math.min(minPos, pos);
      maxPos = Math.max(maxPos, pos);
    }

    int cntWhite = 0;
    int cntBlack = 0;
    int cntGray = 0;

    for (int i = minPos; i <= maxPos; i++) {
      if (currentColor[i + RANGE] == WHITE) {
        cntWhite++;
      } else if (currentColor[i + RANGE] == BLACK) {
        cntBlack++;
      } else if (currentColor[i + RANGE] == GRAY) {
        cntGray++;
      }
    }

    bw.write(String.valueOf(cntWhite));
    bw.write(' ');
    bw.write(String.valueOf(cntBlack));
    bw.write(' ');
    bw.write(String.valueOf(cntGray));

    br.close();
    bw.close();

  }

  private static void paintColor(int color, int target) {
    if (currentColor[target + RANGE] == GRAY) {
      return;
    }

    if (color == WHITE) {
      numWhite[target + RANGE]++;
    } else {
      numBlack[target + RANGE]++;
    }

    if (numWhite[target + RANGE] >= 2 && numBlack[target + RANGE] >= 2) {
      currentColor[target + RANGE] = GRAY;
      return;
    }

    currentColor[target + RANGE] = color;
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter04/lesson03/ChallengePaintingWhiteBlack.kt
package org.dukcode.ps.codetree.trail02.chapter04.lesson03.challengePaintingWhiteBlack

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.StringTokenizer

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

private const val OFFSET = 100_000

private const val NONE = -1
private const val WHITE = 0
private const val BLACK = 1

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/challenge-painting-white-black/description">LINK</a>
 */
fun main() {
    val counts = Array(2) { IntArray(2 * OFFSET + 1) }
    val status = IntArray(2 * OFFSET + 1) { NONE }

    val n = br.readLine().toInt()

    var pos = 0
    repeat(n) {
        val st = StringTokenizer(br.readLine())
        val dist = st.nextToken().toInt()
        val dir = if (st.nextToken() == "L") -1 else 1

        val color = if (dir == -1) WHITE else BLACK

        val nextPos = pos + (dist - 1) * dir

        val start = minOf(pos, nextPos)
        val end = maxOf(pos, nextPos)
        for (idx in start..end) {
            counts[color][idx + OFFSET]++
            status[idx + OFFSET] = color
        }

        pos = nextPos
    }

    val (numWhite, numBlack, numGray) = countColors(counts, status)

    bw.write("$numWhite $numBlack $numGray")

    bw.flush()
}

private fun countColors(
    counts: Array<IntArray>,
    status: IntArray,
): Triple<Int, Int, Int> {
    var numWhite = 0
    var numBlack = 0
    var numGray = 0
    for (idx in status.indices) {
        if (status[idx] == NONE) {
            continue
        }

        if (counts[WHITE][idx] >= 2 && counts[BLACK][idx] >= 2) {
            numGray++
            continue
        }

        if (status[idx] == WHITE) {
            numWhite++
        } else {
            numBlack++
        }
    }
    return Triple(numWhite, numBlack, numGray)
}
